### PR TITLE
Add base objects for Tablet and TabletServer for the upcoming ScanServer feature

### PR DESCRIPTION
**Pull request by dlmarion**
This commit introduces TabletBase, a base class for Tablets, and TabletHostingServer,
an interface for server components that host Tablets. These changes will be used by
the ScanServer feature that is in the works.

Related to #2411

### CODE REVIEW

**Review by ctubbsii — COMMENTED** [review]
Nice. I love incremental changes like this. Even if I can't see the full outcome, it is easy to review, and easy to understand how the code organization can help with the future changes.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletBase.java [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+import org.apache.accumulo.tserver.TabletServerResourceManager;
+import org.apache.accumulo.tserver.metrics.TabletServerScanMetrics;
+import org.apache.accumulo.tserver.scan.ScanParameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This class exists to share code for scanning a tablet between TabletHostingServer implementations
+ */
+public abstract class TabletBase {
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletBase.java:67]: I believe that is the case. That should be the case. @keith-turner wrote the original change against the Scan Server feature branch. I pulled out what I could as a separate PR to get the smaller changeset upstream of our feature. Like you said above, smaller incremental changes that can be reviewed easier.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java [hunk 2]
```diff
@@ -182,6 +169,7 @@ private enum CloseState {
 
   private boolean updatingFlushID = false;
 
+  // TODO volatile ?
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/Tablet.java]: resolved in 58b0e73

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletBase.java [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+
+  private static final Logger log = LoggerFactory.getLogger(TabletBase.class);
+
+  private static final byte[] EMPTY_BYTES = new byte[0];
+
+  protected final KeyExtent extent;
+  protected final ServerContext context;
+
+  // TODO these are written in synch block but read w/o sync so maybe should be volatile
```
[the remark this answers is not in the document]
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletBase.java]: resolved in 58b0e73

**Review by keith-turner — APPROVED** [review]

**Review by ctubbsii — CHANGES_REQUESTED** [review]
The spotbugs checks are failing, but once that is fixed, this is fine.

**Comment by dlmarion** [comment]
Yeah, fixing it now

File: server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletBase.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (keith-turner) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/tablet/TabletBase.java:67]: This change is to isolate the read the read code from the tablet and move it up a level.  In Dave's scan server branch we have a SnapshotTablet that extends TabletBase and offers a readonly view of a snapshot of a tablet.

File: server/tserver/src/main/java/org/apache/accumulo/tserver/TabletHostingServer.java [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+import org.apache.accumulo.fate.zookeeper.ZooCache;
+import org.apache.accumulo.server.GarbageCollectionLogger;
+import org.apache.accumulo.server.ServerContext;
+import org.apache.accumulo.server.conf.TableConfiguration;
+import org.apache.accumulo.tserver.metrics.TabletServerScanMetrics;
+import org.apache.accumulo.tserver.session.Session;
+import org.apache.accumulo.tserver.session.SessionManager;
+import org.apache.accumulo.tserver.tablet.Tablet;
+
+public interface TabletHostingServer {
```
[the remark this answers is not in the document]
Comment (milleruntime) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletHostingServer.java:33]: Can you give a description of this class? I like to put javadoc on interfaces.

[the remark this answers is not in the document]
Comment (dlmarion) [reply server/tserver/src/main/java/org/apache/accumulo/tserver/TabletHostingServer.java:33]: Javadoc for this was added in #2665

**Pull request merged**